src/cls/rbd/cls_rbd.cc

```diff
@@ -4421,6 +4431,337 @@ int group_dir_remove(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
   return 0;
 }
 
+/**
+ * Add an image to the consistency group.
+ *
+ * Input:
+ * @param image_id (std::string)
+ * @param pool_id (int64_t)
+ *
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int group_image_add(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_image_add");
+
+  std::string image_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(image_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  string image_key = RBD_IMAGE_KEY_PREFIX + image_id + "_" + stringify(pool_id);
+
+  bufferlist image_val_bl;
+  int64_t link_state = LINK_DIRTY;
+  ::encode(link_state, image_val_bl);
+  int r = cls_cxx_map_set_val(hctx, image_key, &image_val_bl);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+/**
+ * Remove reference to an image from the consistency group.
+ *
+ * Input:
+ * @param image_id (std::string)
+ * @param pool_id (int64_t)
+ *
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int group_image_remove(cls_method_context_t hctx,
+                       bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_image_remove");
+  std::string image_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(image_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  string image_key = RBD_IMAGE_KEY_PREFIX + image_id + "_" + stringify(pool_id);
+
+  int r = cls_cxx_map_remove_key(hctx, image_key);
+  if (r < 0) {
+    CLS_ERR("error removing image from group: %s", cpp_strerror(r).c_str());
+    return r;
+  }
+
+  return 0;
+}
+
+/**
+ * List images pertaining to the consistency group.
+ *
+ * Input:
+ * none
+ *
+ * Output:
+ * @param length of the list of images
+ * @param tuples of descriptions of the images: image_id, pool_id, image reference state.
+ * @return 0 on success, negative error code on failure
+ */
+int group_image_list(cls_method_context_t hctx,
+                     bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_image_list");
+
+  int max_read = RBD_MAX_KEYS_READ;
+  std::map<string, bufferlist> vals;
+  string last_read = RBD_SNAP_KEY_PREFIX;
+  int r;
+  do {
+    r = cls_cxx_map_get_vals(hctx, RBD_IMAGE_KEY_PREFIX, RBD_IMAGE_KEY_PREFIX,
+			     max_read, &vals);
+    if (r < 0)
+      return r;
+
+    int64_t length = vals.size();
+    ::encode(length, *out);
```

Comment:
In a ```while``` loop that might repeat -- cannot encode the output length here.  Use a standard container and encode everything to the output after exiting the loop.